zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -896,3 +901,264 @@ impl<F: Field> MulAddWordsGadget<F> {
+    // higher `64 - n` bits
+    a64s_hi: [Cell<F>; 4],
+    // shift[0] / 64
+    shf_div64: Cell<F>,
+    // shift[0] % 64
+    shf_mod64: Cell<F>,
+    // 1 << shf_mod64
+    p_lo: Cell<F>,
+    // 1 << (64 - shf_mod64))
+    p_hi: Cell<F>,
+    // shift < 256
+    shf_lt256: IsZeroGadget<F>,
+    // shf_div64 == 0
+    shf_div64_eq0: IsZeroGadget<F>,
+    // shf_div64 == 1
+    shf_div64_eq1: IsEqualGadget<F>,
+    // shf_div64 == 2
+    shf_div64_eq2: IsEqualGadget<F>,
+    // a64s_lo[idx] < p_lo
+    a64s_lo_lt_p_lo: [LtGadget<F, 16>; 4],
+}
+
+impl<F: Field> ShrWordsGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        a: util::Word<F>,
+        shift: util::Word<F>,
+    ) -> Self {
+        let b = cb.query_word();
+        let a64s = array_init(|_| cb.query_cell());
+        let b64s = array_init(|_| cb.query_cell());
+        let a64s_lo = array_init(|_| cb.query_cell());
+        let a64s_hi = array_init(|_| cb.query_cell());
+        let shf_div64 = cb.query_cell();
+        let shf_mod64 = cb.query_cell();
+        let p_lo = cb.query_cell();
+        let p_hi = cb.query_cell();
+        let shf_lt256 = IsZeroGadget::construct(cb, sum::expr(&shift.cells[1..32]));
+        for idx in 0..4 {
+            let offset = idx * N_BYTES_U64;
+
+            // a64s constraint
+            cb.require_equal(
+                "a64s[idx] == from_bytes(a[8 * idx..8 * (idx + 1)])",
+                a64s[idx].expr(),
+                from_bytes::expr(&a.cells[offset..offset + N_BYTES_U64]),
+            );
+
+            // b64s constraint
+            cb.require_equal(
+                "b64s[idx] * shf_lt256 == from_bytes(b[8 * idx..8 * (idx + 1)])",
+                b64s[idx].expr() * shf_lt256.expr(),
+                from_bytes::expr(&b.cells[offset..offset + N_BYTES_U64]),
+            );
+
+            cb.require_equal(
+                "a64s[idx] == a64s_lo[idx] + a64s_hi[idx] * p_lo",
+                a64s[idx].expr(),
+                a64s_lo[idx].expr() + a64s_hi[idx].expr() * p_lo.expr(),
+            );
+        }
+
+        // a64s_lo[idx] < p_lo
+        let a64s_lo_lt_p_lo = array_init(|idx| {
+            let lt = LtGadget::construct(cb, a64s_lo[idx].expr(), p_lo.expr());
+            cb.require_equal("a64s_lo[idx] < p_lo", lt.expr(), 1.expr());
+            lt
+        });
+
+        // merge contraints
+        let shf_div64_eq0 = IsZeroGadget::construct(cb, shf_div64.expr());
+        let shf_div64_eq1 = IsEqualGadget::construct(cb, shf_div64.expr(), 1.expr());
+        let shf_div64_eq2 = IsEqualGadget::construct(cb, shf_div64.expr(), 2.expr());
+        cb.require_equal(
+            "b64s[0] == \
+                (a64s_hi[0] + a64s_lo[1] * p_hi) * shf_div64_eq0 + \
+                (a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq1 + \
+                (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq2 + \
+                a64s_hi[3] * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2)",
```

Comment:
```suggestion
            "Constrain b64s[0]",
```
The current text is too long imo.